forward.go

```diff
@@ -32,64 +34,77 @@ func (s *TunneledConnState) String() string {
 	return out
 }
 
-func (tun *SSHTun) forward(localConn net.Conn) {
-	from := localConn.RemoteAddr().String()
+func (tun *SSHTun) forward(fromConn net.Conn) {
+	from := fromConn.RemoteAddr().String()
 
 	tun.tunneledState(&TunneledConnState{
 		From: from,
 		Info: fmt.Sprintf("accepted %s connection", tun.local.Type()),
 	})
 
-	remoteConn, err := tun.sshClient.Dial(tun.remote.Type(), tun.remote.String())
-	if err != nil {
-		tun.tunneledState(&TunneledConnState{
-			From:  from,
-			Error: fmt.Errorf("remote dial %s to %s failed: %w", tun.remote.Type(), tun.remote.String(), err),
-		})
+	var toConn net.Conn
+	var err error
+	
+	if tun.forwardType == Local {
+		toConn, err = tun.sshClient.Dial(tun.remote.Type(), tun.remote.String())
+		if err != nil {
+			tun.tunneledState(&TunneledConnState{
+				From:  from,
+				Error: fmt.Errorf("remote dial %s to %s failed: %w", tun.remote.Type(), tun.remote.String(), err),
+			})
+
+			fromConn.Close()
+			return
+		}
+	}
+	if tun.forwardType == Remote {
+		toConn, err = net.Dial(tun.local.Type(), tun.local.String())
+		if err != nil {
+			tun.tunneledState(&TunneledConnState{
+				From:  from,
+				Error: fmt.Errorf("remote dial %s to %s failed: %w", tun.local.Type(), tun.local.String(), err),
+			})
 
-		localConn.Close()
-		return
+			fromConn.Close()
+			return
+		}
 	}
 
 	connStr := fmt.Sprintf("%s -(%s)> %s -(ssh)> %s -(%s)> %s", from, tun.local.Type(), tun.local.String(),
 		tun.server.String(), tun.remote.Type(), tun.remote.String())
 
 	tun.tunneledState(&TunneledConnState{
-		From:  from,
-		Info:  fmt.Sprintf("connection stablished: %s", connStr),
-		Ready: true,
+		From:   from,
+		Info:   fmt.Sprintf("connection established: %s", connStr),
+		Ready:  true,
+		Closed: false,
 	})
 
 	connCtx, connCancel := context.WithCancel(tun.ctx)
 	errGroup := &errgroup.Group{}
 
 	errGroup.Go(func() error {
 		defer connCancel()
-		_, err = io.Copy(remoteConn, localConn)
+		_, err = io.Copy(toConn, fromConn)
 		if err != nil {
 			return fmt.Errorf("failed copying bytes from remote to local: %w", err)
 		}
-
-		return nil
+		return toConn.Close()
 	})
 
 	errGroup.Go(func() error {
 		defer connCancel()
-		_, err = io.Copy(localConn, remoteConn)
+		_, err = io.Copy(fromConn, toConn)
 		if err != nil {
 			return fmt.Errorf("failed copying bytes from local to remote: %w", err)
 		}
-
-		return nil
+		return fromConn.Close()
 	})
 
-	<-connCtx.Done()
-
-	localConn.Close()
-	remoteConn.Close()
-
 	err = errGroup.Wait()
```

Comment:
I discovered that if multiple tunnels are chained one after the other, the order of closing connections would make one of the tunnels return an error, but I don't remember which one since I discovered this last year. :)

Reply:
Ok, I see! I think we can investigate that problem in another PR. For now I will merge this PR in an intermediate branch, remove this part and integrate it with the main branch.
Then I will open another one to test this changes regarding connection closing to see if they don't cause other problems to other people using the library.